### PR TITLE
[Tree unique] Substitution rules

### PR DESCRIPTION
**Pull request by rtjoa**
Add rules for the following functions:
```
(function SubstExpr (Expr Expr) Expr)
(function SubstListExpr (ListExpr Expr) ListExpr)
```

## Stack
1. https://github.com/egraphs-good/eggcc/pull/206
2. https://github.com/egraphs-good/eggcc/pull/212
3. https://github.com/egraphs-good/eggcc/pull/213
4. https://github.com/egraphs-good/eggcc/pull/214
5. https://github.com/egraphs-good/eggcc/pull/216

### CODE REVIEW

File: tree_unique_args/src/subst.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    });
+
+    let sort = ctor.sort().name();
+    let br = "\n         ";
+    format!("(rewrite (Subst{sort} {ctor_pattern} v){br}{substed_ctor}{br}:ruleset always-run)")
+}
+
+pub(crate) fn subst_rules() -> Vec<String> {
+    ESort::iter()
+        .map(|sort| "(function Subst* (* Expr) * :unextractable)".replace("*", sort.name()))
```
Comment (oflatt) [inline tree_unique_args/src/subst.rs:30]: use `format!`

**Review by oflatt — APPROVED** [review]
Very nice!

File: tree_unique_args/src/subst.rs (hunk 1 above)
Comment (rtjoa) [reply tree_unique_args/src/subst.rs:30]: That's what I did at first, but I changed to `.replace` so we don't have to bind `sort.name()`. I think this is still readable, we often write `Subst*` when talking about families of functions anyway

Comment (oflatt) [reply tree_unique_args/src/subst.rs:30]: Ok

**Pull request merged**